modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java

```diff
@@ -12051,4 +12055,25 @@ public static long uncompressedSize(File zip) throws IOException {
             return size;
         }
     }
+
+    /**
+     * Acquires the given semaphore, executes the given callable and schedules the release of permits asynchronously
+     */
+    public static <T> IgniteFuture<T> acquireAndExecute(IgniteSemaphore semaphore,
+                                          IgniteCallable<IgniteFuture<T>> callable, int numPermits) throws Exception {
+
+        semaphore.acquire(numPermits);
+
+        IgniteFuture<T> future = callable.call();
+
+        future.listen(new IgniteInClosure<IgniteFuture<T>>() {
+            @Override public void apply(IgniteFuture<T> IgniteFuture) {
+                if (IgniteFuture.isCancelled() || IgniteFuture.isDone()) {
```

Comment:
Not sure if these checks needed here: future's listen callback only executed when future is completed